### PR TITLE
 Support OneOf #38 

### LINKED ISSUE #38: Support OneOf
We should add support to the `OneOf` feature of protobuf. When encoding `OneOf` properties we should only add one leaf to the tree for each field. 

```protobuf,
message SubMessage {
    string name = 1;
    int number = 2;   
}
message SampleMessage {
  oneof test_oneof {
     string name = 4;
     SubMessage sub_message = 9;
  }
}
```

This issue has a gitcoin bounty attached to it. If you'd like to start working on it, feel free to reach out to me (lucas@centrifuge.io) directly, join our slack channel `#bounties` on our [slack](https://join.slack.com/t/centrifuge-io/shared_invite/enQtNDYwMzQ5ODA3ODc0LTU4ZjU0NDNkOTNhMmUwNjI2NmQ2MjRiNzA4MGIwYWViNTkxYzljODU2OTk4NzM4MjhlOTNjMDAwNWZkNzY2YWY)

### PR DESCRIPTION
**Pull request by alexeyneu**
Fixes #38

### CODE REVIEW

File: proofs/flatten.go [hunk 1]
```diff
@@ -55,8 +55,12 @@ func (f *messageFlattener) handleValue(prop Property, value reflect.Value, saltV
 
 		// Handle each field of the struct
 		for i := 0; i < value.NumField(); i++ {
+			var oneof_here bool
 			field := value.Type().Field(i)
-
+			if field.Tag.Get("protobuf_oneof") != "" && !value.Field(i).IsNil(){
+				field = value.Field(i).Elem().Elem().Type().Field(0)
```
[the remark this answers is not in the document]
Comment (alexeyneu) [reply proofs/flatten.go]: It's `oneof` handling in language with no unions. Not that thing you'll catch on quickly.

[the remark this answers is not in the document]
Comment (alexeyneu) [reply proofs/flatten.go]: in `field` you have entire type of currently toggled `oneof` field-set field

[the remark this answers is not in the document]
Comment (alexeyneu) [reply proofs/flatten.go]: https://github.com/golang/protobuf/blob/347cf4a86c1cb8d262994d8ef5924d4576c5b331/jsonpb/jsonpb.go#L323

File: proofs/flatten.go [hunk 2]
```diff
@@ -55,8 +55,12 @@ func (f *messageFlattener) handleValue(prop Property, value reflect.Value, saltV
 
 		// Handle each field of the struct
 		for i := 0; i < value.NumField(); i++ {
+			var oneof_here bool
```
Comment (lucasvo) [inline proofs/flatten.go]: Could you rename this to oneOfField?

File: proofs/salts_test.go [hunk 3]
```diff
@@ -28,7 +28,7 @@ func TestFillSalts(t *testing.T) {
 
 	//Document with oneof fieldset
```
Comment (lucasvo) [inline proofs/salts_test.go]:
```
	// Document with oneof fieldset
```
Space between double dashes and comment please

File: proofs/tree_test.go [hunk 4]
```diff
@@ -1316,6 +1316,38 @@ func TestCreateProof_standard_customEncoder(t *testing.T) {
...
 }
 
+func TestCreateOneofProof(t *testing.T) {
+	doctree := NewDocumentTree(TreeOptions{Hash: sha256Hash})
+	err := doctree.AddLeavesFromDocument(&documentspb.ExampleOneofSampleDocument, &documentspb.ExampleOneofSampleSaltsDocument)
+	assert.Nil(t, err)
+	err = doctree.Generate()
+
+	_, err = doctree.CreateProof("InexistentField")
+	assert.EqualError(t, err, "No such field: InexistentField in obj")
```
Comment (lucasvo) [inline proofs/tree_test.go]: You probably don't need to test this here if it's tested elsewhere.

File: proofs/tree_test.go [hunk 5]
```diff
@@ -1316,6 +1316,38 @@ func TestCreateProof_standard_customEncoder(t *testing.T) {
...
+	assert.Equal(t, rootHash, doctree.rootHash)
+	valid, err := ValidateProofHashes(fieldHash, proof.Hashes, rootHash, doctree.hash)
+	assert.True(t, valid)
+
+	valid, err = doctree.ValidateProof(&proof)
+	assert.True(t, valid)
+	assert.Nil(t, err)
+
+	falseProof, err := doctree.CreateProof("valueB")
+	falseProof.Value = "Invalid"
```
Comment (lucasvo) [inline proofs/tree_test.go]: this test doesn't test much and is covered in other test cases already

File: proofs/tree_test.go [hunk 6]
```diff
@@ -1316,6 +1316,38 @@ func TestCreateProof_standard_customEncoder(t *testing.T) {
...
+	err := doctree.AddLeavesFromDocument(&documentspb.ExampleOneofSampleDocument, &documentspb.ExampleOneofSampleSaltsDocument)
+	assert.Nil(t, err)
+	err = doctree.Generate()
+
+	_, err = doctree.CreateProof("InexistentField")
+	assert.EqualError(t, err, "No such field: InexistentField in obj")
+
+	proof, err := doctree.CreateProof("valueB")
+	assert.Nil(t, err)
+	assert.Equal(t, ReadableName("valueB"), proof.Property)
```
Comment (lucasvo) [inline proofs/tree_test.go]: Could you create a test for the following additional scenario:  Try to create a proof for valueC when valueB is set

Comment (lucasvo) [reply proofs/tree_test.go]: And when both are nil create a proof for either one (or even both) when both are nil.

**Comment by lucasvo** [comment]
We've changed how salts are handled in #43 and that code has now been merged. Could you resolve the conflicts please?

**Comment by alexeyneu** [comment]
it's coz scott agreed to pay part of money himself

**Comment by lucasvo** [comment]
@alexeyneu thank you for resolving the conflicts. @vimukthi-git will review the PR and test it locally. We might still have some smaller feedback but given that is all addressed we'll merge it and pay out the bounty.

**Pull request merged**